### ISSUE TITLE
Opacity for edges

### ISSUE THREAD

**Issue by h-a-s-k** [jacomyal/sigma.js#1426]
When exporting using the gephi plugin, opacity seems to work fine, however with the latest version of sigma.js it doesn't seem to work when setting the edge color to a rgba value. This would be nice to have since sigma.js has great performance (and doesn't leak memory like a few others) but not having opacity support is a dealbreaker for me.

**Comment by jacomyal** [comment]
Hum, that's weird, sigma does support alpha values:
![image](https://github.com/jacomyal/sigma.js/assets/144846/2f50fe52-38dd-4406-8c11-e57930bbf871)

It looks like there is an issue with how the alpha channel is handled though, because the whiter nodes and edges are, the less the alpha value is considered apparently. So I'll leave the issue opened until I solve this bug.

@h-a-s-k Could you please send some data and/or code, to help me reproduce your issue? Thanks!

**Comment by RaphaelHansen** [comment]
Hi Alexis, 

In case h-a-s-k doesn't get back to you.

My experience is that the 'whiter' or 'brighter' a colour is the more likely it is to turn an opaque white on overlap. 

`    graph.forEachEdge((edge, attributes, source, target) => {
      const targetColor = graph.getNodeAttribute(target, "color") || "#ccc";
      const edgeOpacity = 0.5; 
      
      if (targetColor.startsWith("rgb(")) {
        const rgbValues = targetColor.substring(4, targetColor.length - 1).split(",");
        const edgeColor = `rgba(${rgbValues[0]}, ${rgbValues[1]}, ${rgbValues[2]}, ${edgeOpacity})`;
        graph.setEdgeAttribute(edge, "color", edgeColor);
      } else {
        const rgbColor = hexToRgb(targetColor);
        if (rgbColor) {
          const edgeColor = `rgba(${rgbColor.r}, ${rgbColor.g}, ${rgbColor.b}, ${edgeOpacity})`;
          graph.setEdgeAttribute(edge, "color", edgeColor);
        } else {
          graph.setEdgeAttribute(edge, "color", targetColor);
        }
      }
    });`


Here you can see that the lighter pink/purple colour turns white on overlap.
![Screenshot 2024-05-28 at 11 22 57 am](https://github.com/jacomyal/sigma.js/assets/156052991/4463bda6-7e08-4dbb-8c94-7c6c8bf5f09c)

Darker colours also get brighter, but they are able to avoid turning completely white:
![Screenshot 2024-05-28 at 11 18 22 am](https://github.com/jacomyal/sigma.js/assets/156052991/4b22e669-88c9-4ca6-9e18-9a576241206e)

I'm no expert in this field but my hypothesis is that it seems that the rgb values are also compounding ontop of eachother on overlap rather than the expected outcome of only the alpha value increasing when edges overlap. 

Hopefully this makes sense.

**Comment by h-a-s-k** [comment]
I tried setting the alpha both as @RaphaelHansen did and also like you showed in https://github.com/jacomyal/sigma.js/commit/68da672a69e80cc8d1ced50d13f6268981d89809. On my graph with 120k+ edges using sigma@3.0.0-beta.19.

When every node and edge is set to red with an alpha of 0.01

```js
edges.forEach((edge, idx) => {
	const addedEdge = graph.addEdge(edge.source, edge.target, { size: 1 });
	graph.setEdgeAttribute(addedEdge, `color`, chroma('#ff0000').alpha(0.01).hex());
});
```

looks like this:
![sigma](https://github.com/jacomyal/sigma.js/assets/140975143/1f7521c9-5d09-4188-a083-e8cc03dabaf5)


compared to pixi.js with same color and alpha:
![pixi](https://github.com/jacomyal/sigma.js/assets/140975143/75cb4d78-29c8-4444-8891-39a8fb1c402f)


Also could be placebo but when panning around using the reducer method it felt slower than setting a color directly with setEdgeAttribute once:
```js
const instance = new Sigma(graph, canvasContainer, {
	labelThreshold: 10,
	hideEdgesOnMove: false,
	renderEdgeLabels: false,

	edgeReducer: (_, attr) => {
		attr.color = chroma('#ff0000').alpha(0.01).hex();
		return attr;
	},
});
```

**Comment by ck-coatue** [comment]
I've noticed that EdgeCurvedArrowProgram handles opacity well (doesn't wash out like the other edge/node programs).

**Comment by h-a-s-k** [comment]
> I've noticed that EdgeCurvedArrowProgram handles opacity well (doesn't wash out like the other edge/node programs).

Definitely an improvement

![test sigma](https://github.com/jacomyal/sigma.js/assets/140975143/1b3c9f01-c87e-4deb-9d49-3a5294800fc3)

**Comment by ck-coatue** [comment]
figured out if u want to do the same for nodes as well, clone the NodeCircleProgram (as example) and change line:

https://github.com/jacomyal/sigma.js/blob/23e53736c9a95bd0bc81bdef9644be11e6d630c2/packages/sigma/src/rendering/programs/node-circle/frag.glsl.ts#L31

from

```
gl_FragColor = mix(v_color, transparent, t);
```

to

```
gl_FragColor = mix(vec4(v_color.rgb * v_color.a, v_color.a), transparent, t);
```

I took inspiration from https://github.com/jacomyal/sigma.js/blob/23e53736c9a95bd0bc81bdef9644be11e6d630c2/packages/edge-curve/src/shader-frag.ts#L68

**Comment by benny-noumena** [comment]
@ck-coatue It seems that this has be undone in the current beta relase. 

https://github.com/jacomyal/sigma.js/blob/721ac7e4dd02e00e828cd1538dc2310ed446809a/packages/edge-curve/src/shader-frag.ts#L73

Maybe it was for performance reason but I would love to have a way to make edges semi-transparent.

**Comment by h-a-s-k** [comment]
> @ck-coatue It seems that this has be undone in the current beta relase.
> 
> https://github.com/jacomyal/sigma.js/blob/721ac7e4dd02e00e828cd1538dc2310ed446809a/packages/edge-curve/src/shader-frag.ts#L73
> 
> Maybe it was for performance reason but I would love to have a way to make edges semi-transparent.

Yeah last version with working transparency is `@sigma/edge-curve@3.0.0-beta.10` for me

**Comment by seoanezonjic** [comment]
Any update about this @jacomyal ? I've the same problem that @h-a-s-k and my larger graphs looks like he shown:
 ![sigma](https://github.com/jacomyal/sigma.js/assets/140975143/1f7521c9-5d09-4188-a083-e8cc03dabaf5)

Thank you in advance
Pedro Seoane

**Comment by jacomyal** [comment]
Hi @seoanezonjic,

So, basically, I want to address this issue with other opacity and z-indexing related issues (#1233 and #1427), by implementing [Order Independent Transparency](https://en.wikipedia.org/wiki/Order-independent_transparency), and joining edges and nodes on the same unique WebGL layer.

This is a tough and impactful feature, which will be the core of [version 4](https://github.com/jacomyal/sigma.js/milestone/11) in some future. Unfortunately, our open-source resources at @ouestware are focused on [Gephi Lite](https://github.com/gephi/gephi-lite/) at the moment. I hope to be able to get some planed scheduled for v4 by the end of the year.

I hope this helps!

**Comment by jacomyal** [comment]
Some follow-up on this issue:
- We just released a [v4-alpha version](https://github.com/jacomyal/sigma.js/discussions/1539)
- In this version, nodes and edges all appear in the same WebGL layer
- In this version, nodes and edges opacity [works well](https://v4.sigmajs.org/how-to/nodes/colors-sizes/#opacity-and-alpha-blending)

I am cleaning the issues now, to have a better overview of what's left to solve, so I'm closing this ticket.
